File: src/lib/api/copilotByok.ts
import { invoke } from "@tauri-apps/api/core";
import type { UsageScript } from "@/types";
import type { StreamCheckResult } from "./connectivity-check";

export type CopilotByokApiType = "chat-completions" | "responses" | "messages";

export type CopilotByokReasoningEffortFormat = CopilotByokApiType;

export type CopilotByokEditTool =
  | "find-replace"
  | "multi-find-replace"
  | "apply-patch"
  | "code-rewrite";

export interface CopilotByokModel {
  id: string;
  modelId: string;
  name: string;
  enabled: boolean;
  toolCalling?: boolean | null;
  vision?: boolean | null;
  thinking?: boolean | null;
  streaming?: boolean | null;
  contextWindow?: number | null;
  maxInputTokens?: number | null;
  maxOutputTokens?: number | null;
  editTools: string[];
  zeroDataRetentionEnabled: boolean;
  supportsReasoningEffort: string[];
  reasoningEffortFormat: CopilotByokReasoningEffortFormat | null;
  modelOptions: unknown;
  extra: Record<string, unknown>;
}

export interface CopilotByokGroup {
  id: string;
  name: string;
  url: string;
  apiKey: string;
  apiType: CopilotByokApiType;
  websiteUrl?: string | null;
  notes?: string | null;
  icon?: string | null;
  iconColor?: string | null;
  category?: "official" | "custom" | null;
  usageScript?: UsageScript | null;
  enabled: boolean;
  requestHeaders: Record<string, string>;
  models: CopilotByokModel[];
  extra: Record<string, unknown>;
}

export type VsCodeEdition = "stable" | "insiders";

export interface CopilotByokTargetState {
  id: string;
  source: "detected" | "custom";
  edition: VsCodeEdition | null;
  editionName: string | null;
  profileId: string | null;
  profileName: string;
  isDefault: boolean;
  languageModelsPath: string;
  configExists: boolean;
  backupExists: boolean;
  selected: boolean;
  managedGroupCount: number;
  readError: string | null;
}

export interface CopilotCliState {
  supported: boolean;
  enabled: boolean;
  selectedGroupId: string | null;
  selectedModelId: string | null;
  selectedProviderName: string | null;
  selectedModelName: string | null;
  environmentMatches: boolean;
  environmentConflicts: string[];
  officialActivationRequiresConfirmation: boolean;
}

export interface CopilotByokState {
  groups: CopilotByokGroup[];
  targets: CopilotByokTargetState[];
  selectedTargetIds: string[];
  managedModelCount: number;
  cli: CopilotCliState;
}

export interface CopilotByokSyncResult {
  targetIds: string[];
  managedModelCount: number;
  changedTargetCount: number;
}

export interface CopilotByokImportResult {
  targetId: string;
  importedGroupCount: number;
  importedModelCount: number;
  reusedModelCount: number;
  skippedGroupCount: number;
  changedTargetCount: number;
  warnings: string[];
}

export const copilotByokApi = {
  getState(): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_get_state");
  },

  setCliSelection(groupId: string, modelId: string): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_set_cli_selection", {
      groupId,
      modelId,
    });
  },

  disableCli(): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_disable_cli");
  },

  setTargets(targetIds: string[]): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_set_targets", {
      targetIds,
    });
  },

  addCustomTarget(
    path: string,
    name?: string | null,
  ): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_add_custom_target", {
      path,
      name: name || null,
    });
  },

  removeCustomTarget(targetId: string): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_remove_custom_target", {
      targetId,
    });
  },

  upsertGroup(group: CopilotByokGroup): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_upsert_group", {
      group,
    });
  },

  deleteGroup(groupId: string): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_delete_group", {
      groupId,
    });
  },

  reorderGroups(groupIds: string[]): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_reorder_groups", {
      groupIds,
    });
  },

  importModels(targetId: string): Promise<CopilotByokImportResult> {
    return invoke<CopilotByokImportResult>("copilot_byok_import_models", {
      targetId,
    });
  },

  sync(): Promise<CopilotByokSyncResult> {
    return invoke<CopilotByokSyncResult>("copilot_byok_sync");
  },

  restoreBackup(targetId: string): Promise<boolean> {
    return invoke<boolean>("copilot_byok_restore_backup", { targetId });
  },

  checkConnection(groupId: string): Promise<StreamCheckResult> {
    return invoke<StreamCheckResult>("copilot_byok_check_connection", {
      groupId,
    });
  },

  updateUsageScript(
    groupId: string,
    usageScript: UsageScript,
  ): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_byok_update_usage_script", {
      groupId,
      usageScript,
    });
  },
};

/** Independent first-class GitHub Copilot CLI provider catalog and switcher. */
export const copilotCliApi = {
  getState(): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_cli_get_state");
  },

  setSelection(
    groupId: string,
    groupName: string,
    confirmUnmanagedClear = false,
  ): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_cli_set_selection", {
      groupId,
      groupName,
      confirmUnmanagedClear,
    });
  },

  disable(): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_cli_disable");
  },

  upsertGroup(group: CopilotByokGroup): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_cli_upsert_group", { group });
  },

  deleteGroup(groupId: string): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_cli_delete_group", { groupId });
  },

  reorderGroups(groupIds: string[]): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_cli_reorder_groups", { groupIds });
  },

  checkConnection(groupId: string): Promise<StreamCheckResult> {
    return invoke<StreamCheckResult>("copilot_cli_check_connection", {
      groupId,
    });
  },

  updateUsageScript(
    groupId: string,
    usageScript: UsageScript,
  ): Promise<CopilotByokState> {
    return invoke<CopilotByokState>("copilot_cli_update_usage_script", {
      groupId,
      usageScript,
    });
  },

  openTerminal(groupId: string, cwd: string): Promise<boolean> {
    return invoke<boolean>("copilot_cli_open_terminal", { groupId, cwd });
  },
};

File: src-tauri/src/commands/copilot_byok.rs
use crate::app_config::AppType;
use crate::copilot_byok::{
    self, CopilotByokGroup, CopilotByokImportResult, CopilotByokState, CopilotByokSyncResult,
};
use crate::provider::UsageScript;
use crate::services::stream_check::{StreamCheckResult, StreamCheckService};
use crate::services::{McpService, PromptService};
use crate::store::AppState;
use tauri::{AppHandle, Emitter, State};

fn finish_cli_mutation(
    app: &AppHandle,
    result: Result<CopilotByokState, crate::error::AppError>,
) -> Result<CopilotByokState, String> {
    let next = result.map_err(String::from)?;
    crate::tray::refresh_tray_menu(app);
    if let Err(error) = app.emit("copilot-cli-state-changed", &next) {
        log::warn!("Failed to emit Copilot CLI state change: {error}");
    }
    Ok(next)
}

fn reproject_shared_resources(state: &AppState) -> Result<(), String> {
    McpService::sync_enabled_for_app(state, &AppType::CopilotByok).map_err(String::from)?;

    let prompts = PromptService::get_prompts(state, AppType::CopilotByok).map_err(String::from)?;
    if let Some(prompt) = prompts.values().find(|prompt| prompt.enabled) {
        PromptService::upsert_prompt(state, AppType::CopilotByok, &prompt.id, prompt.clone())
            .map_err(String::from)?;
    }

    Ok(())
}

#[tauri::command]
pub fn copilot_byok_get_state(state: State<'_, AppState>) -> Result<CopilotByokState, String> {
    copilot_byok::get_state(state.db.as_ref()).map_err(Into::into)
}

#[tauri::command]
pub fn copilot_cli_get_state(state: State<'_, AppState>) -> Result<CopilotByokState, String> {
    copilot_byok::get_cli_state(state.db.as_ref()).map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_set_cli_selection(
    state: State<'_, AppState>,
    group_id: String,
    model_id: String,
) -> Result<CopilotByokState, String> {
    copilot_byok::set_cli_selection(state.db.as_ref(), &group_id, &model_id).map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_cli_set_selection(
    app: AppHandle,
    state: State<'_, AppState>,
    group_id: String,
    group_name: Option<String>,
    confirm_unmanaged_clear: Option<bool>,
) -> Result<CopilotByokState, String> {
    let result = copilot_byok::set_cli_provider(
        state.db.as_ref(),
        &group_id,
        group_name.as_deref(),
        confirm_unmanaged_clear.unwrap_or(false),
    );
    let result = result.map_err(|error| {
        log::error!(
            "Failed to activate Copilot CLI provider id='{}' name='{}': {}",
            group_id,
            group_name.as_deref().unwrap_or(""),
            error
        );
        error
    });
    finish_cli_mutation(&app, result)
}

#[tauri::command]
pub fn copilot_byok_disable_cli(state: State<'_, AppState>) -> Result<CopilotByokState, String> {
    copilot_byok::disable_cli(state.db.as_ref()).map_err(Into::into)
}

#[tauri::command]
pub fn copilot_cli_disable(
    app: AppHandle,
    state: State<'_, AppState>,
) -> Result<CopilotByokState, String> {
    finish_cli_mutation(&app, copilot_byok::disable_cli(state.db.as_ref()))
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_cli_open_terminal(
    state: State<'_, AppState>,
    group_id: String,
    cwd: Option<String>,
) -> Result<bool, String> {
    let environment =
        copilot_byok::cli_launch_environment(state.db.as_ref(), &group_id).map_err(String::from)?;
    super::misc::launch_terminal_running_with_env("copilot", "copilot_cli", &environment, cwd)?;
    Ok(true)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_update_usage_script(
    state: State<'_, AppState>,
    group_id: String,
    usage_script: UsageScript,
) -> Result<CopilotByokState, String> {
    copilot_byok::update_usage_script(state.db.as_ref(), &group_id, usage_script)
        .map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_cli_update_usage_script(
    app: AppHandle,
    state: State<'_, AppState>,
    group_id: String,
    usage_script: UsageScript,
) -> Result<CopilotByokState, String> {
    finish_cli_mutation(
        &app,
        copilot_byok::update_cli_usage_script(state.db.as_ref(), &group_id, usage_script),
    )
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_set_targets(
    state: State<'_, AppState>,
    target_ids: Vec<String>,
) -> Result<CopilotByokState, String> {
    let updated = copilot_byok::set_targets(state.db.as_ref(), target_ids).map_err(String::from)?;
    reproject_shared_resources(state.inner())?;
    Ok(updated)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_add_custom_target(
    state: State<'_, AppState>,
    path: String,
    name: Option<String>,
) -> Result<CopilotByokState, String> {
    let updated =
        copilot_byok::add_custom_target(state.db.as_ref(), path, name).map_err(String::from)?;
    reproject_shared_resources(state.inner())?;
    Ok(updated)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_remove_custom_target(
    state: State<'_, AppState>,
    target_id: String,
) -> Result<CopilotByokState, String> {
    let updated =
        copilot_byok::remove_custom_target(state.db.as_ref(), &target_id).map_err(String::from)?;
    reproject_shared_resources(state.inner())?;
    Ok(updated)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_upsert_group(
    state: State<'_, AppState>,
    group: CopilotByokGroup,
) -> Result<CopilotByokState, String> {
    copilot_byok::upsert_group(state.db.as_ref(), group).map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_delete_group(
    state: State<'_, AppState>,
    group_id: String,
) -> Result<CopilotByokState, String> {
    copilot_byok::delete_group(state.db.as_ref(), &group_id).map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_reorder_groups(
    state: State<'_, AppState>,
    group_ids: Vec<String>,
) -> Result<CopilotByokState, String> {
    copilot_byok::reorder_groups(state.db.as_ref(), group_ids).map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_cli_upsert_group(
    app: AppHandle,
    state: State<'_, AppState>,
    group: CopilotByokGroup,
) -> Result<CopilotByokState, String> {
    finish_cli_mutation(
        &app,
        copilot_byok::upsert_cli_group(state.db.as_ref(), group),
    )
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_cli_delete_group(
    app: AppHandle,
    state: State<'_, AppState>,
    group_id: String,
) -> Result<CopilotByokState, String> {
    finish_cli_mutation(
        &app,
        copilot_byok::delete_cli_group(state.db.as_ref(), &group_id),
    )
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_cli_reorder_groups(
    app: AppHandle,
    state: State<'_, AppState>,
    group_ids: Vec<String>,
) -> Result<CopilotByokState, String> {
    finish_cli_mutation(
        &app,
        copilot_byok::reorder_cli_groups(state.db.as_ref(), group_ids),
    )
}

#[tauri::command]
pub fn copilot_byok_sync(state: State<'_, AppState>) -> Result<CopilotByokSyncResult, String> {
    copilot_byok::sync(state.db.as_ref()).map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_import_models(
    state: State<'_, AppState>,
    target_id: String,
) -> Result<CopilotByokImportResult, String> {
    copilot_byok::import_models(state.db.as_ref(), &target_id).map_err(Into::into)
}

#[tauri::command(rename_all = "camelCase")]
pub fn copilot_byok_restore_backup(
    state: State<'_, AppState>,
    target_id: String,
) -> Result<bool, String> {
    copilot_byok::restore_backup(state.db.as_ref(), &target_id).map_err(Into::into)
}

/// 只检查供应商 Base URL 是否可达，不调用模型列表或真实推理接口。
#[tauri::command(rename_all = "camelCase")]
pub async fn copilot_byok_check_connection(
    state: State<'_, AppState>,
    group_id: String,
) -> Result<StreamCheckResult, String> {
    let group = copilot_byok::usage_catalog(state.db.as_ref())
        .map_err(String::from)?
        .into_iter()
        .find(|group| group.id == group_id)
        .ok_or_else(|| format!("VS Code Copilot provider {group_id} does not exist"))?;
    let config = state.db.get_stream_check_config().map_err(String::from)?;
    let result = StreamCheckService::check_url_with_retry(&group.url, &config)
        .await
        .map_err(String::from)?;
    let _ = state
        .db
        .save_stream_check_log(&group.id, &group.name, "copilot-byok", &result);
    Ok(result)
}

#[tauri::command(rename_all = "camelCase")]
pub async fn copilot_cli_check_connection(
    state: State<'_, AppState>,
    group_id: String,
) -> Result<StreamCheckResult, String> {
    let group = copilot_byok::cli_usage_catalog(state.db.as_ref())
        .map_err(String::from)?
        .into_iter()
        .find(|group| group.id == group_id)
        .ok_or_else(|| format!("Copilot CLI provider {group_id} does not exist"))?;
    let config = state.db.get_stream_check_config().map_err(String::from)?;
    let result = StreamCheckService::check_url_with_retry(&group.url, &config)
        .await
        .map_err(String::from)?;
    let _ = state
        .db
        .save_stream_check_log(&group.id, &group.name, "copilot-cli", &result);
    Ok(result)
}
